Merge 1.6.15

iMichka opened the pull request.

mistydemeo committed:
Xcode: fix version detection with unknown clang

sjackman committed:
Relocate bottles on Linux using patchelf

Ensure patchelf is installed to pour bottles and build bottles.

maxim-belkin committed:
Properly append values to HOMEBREW_CCCFG

maxim-belkin committed:
Add make shim to Linux super env

maxim-belkin committed:
Remove infinite loophole

reitermarkus committed:
Separate staging from download.

commitay committed:
audit: ensure `bottle` is defined

reitermarkus committed:
Allow unused keyword arguments.

reitermarkus committed:
Fix `basename_without_params`.

reitermarkus committed:
Refactor unpack strategy specs.

reitermarkus committed:
Fix `XzUnpackStrategy`.

reitermarkus committed:
Reuse `UncompressedUnpackStrategy#extract_to_dir`.

reitermarkus committed:
Add `@ref_type` and `@ref` needef for `fossil`.

reitermarkus committed:
`xar` is only pre-installed on macOS.

reitermarkus committed:
Use more descriptive heredoc names.

commitay committed:
audit: check for optional and recommended requirements on new formulae

MikeMcQuaid committed:
Merge pull request #4455 from reitermarkus/heredocs

Use more descriptive heredoc names.

MikeMcQuaid committed:
Merge pull request #4281 from mistydemeo/fix_unparsed_clang_exception

Xcode: fix version detection with unknown clang

maxim-belkin committed:
extend/ENV/shared.rb: [new method] append_to_cccfg

append_to_cccfg properly appends values to HOMEBREW_CCCFG variable

MikeMcQuaid committed:
Merge pull request #4381 from maxim-belkin/make-shim-linux

Add make shim to Linux super env

MikeMcQuaid committed:
shared: fix HOMEBREW_CCCFG reference.

reitermarkus committed:
Merge pull request #4402 from reitermarkus/separate-staging

Separate staging from download.

MikeMcQuaid committed:
Merge pull request #4380 from maxim-belkin/fix-cccfgfgccfg

Properly append values to HOMEBREW_CCCFG

MikeMcQuaid committed:
Merge pull request #4382 from sjackman/keg_relocate

Relocate bottles on Linux using patchelf

DomT4 committed:
hardware: add dodeca-core for 2018 MBPs

MikeMcQuaid committed:
Use Ruby 2.3.7

Fixes #4459.

DomT4 committed:
xcode: update expected Clang on 10.14

commitay committed:
Merge pull request #4420 from commitay/bottle

audit: ensure `bottle` is defined

commitay committed:
Merge pull request #4456 from commitay/optional-reqs

audit: check for optional and recommended requirements on new formulae

MikeMcQuaid committed:
extend/fileutils: fix monkeypatch.

Otherwise using the options argument or relying on a result means this
method will fail. This is why monkeypatching sucks.

CC @reitermarkus as we've talked about monkeypatching recently.

MikeMcQuaid committed:
extend/os/linux/hardware/cpu: return :dunno more often.

This matches the behaviour described in the test and on macOS.

MikeMcQuaid committed:
test/cask/cli_spec: temporary skip some tests.

These seem to be related to a leaking RSpec double between tests and
I've verified `brew cask install` still behaves as expected so
let's skip them temporarily.

MikeMcQuaid committed:
Merge pull request #4461 from MikeMcQuaid/portable-ruby-2.3.7

Use Ruby 2.3.7

MikeMcQuaid committed:
Merge pull request #4462 from DomT4/clang

xcode: update expected Clang on 10.14

MikeMcQuaid committed:
Merge pull request #4460 from DomT4/cores

hardware: add dodeca-core for 2018 MBPs

MikeMcQuaid committed:
ruby.sh: cleanup gems on portable Ruby installation.

This avoids `brew style` and friends from getting upset when they
attempt to use the native gem extensions from a previous portable Ruby
or a system Ruby.

ABI? What's that?

MikeMcQuaid committed:
Merge pull request #4467 from MikeMcQuaid/portable-ruby-cleanup-gems

ruby.sh: cleanup gems on portable Ruby installation.

iMichka committed:
Merge tag Homebrew/1.6.15 into Linuxbrew/master

 Conflicts:
	Library/Homebrew/cmd/vendor-install.sh
	Library/Homebrew/dev-cmd/bottle.rb
	Library/Homebrew/extend/os/linux/keg_relocate.rb
	Library/Homebrew/formula_installer.rb
	Library/Homebrew/test/cmd/install_spec.rb

sjackman commented:
Homebrew/core does not have a `x86_64_linux` bottle for `patchelf`, so this bug does not occur when testing with Homebrew/brew.
At first glance the bug is that installing a bottle requires `patchelf`, so it tries to install `patchelf` from a bottle, which requires `patchelf`, et c.

sjackman commented:
That bug does not exist in the previous Linuxbrew/brew version of this code, but appears to exist in the code that I sent upstream to Homebrew/brew.

iMichka commented:
There is also something weird with the portable ruby download:
```
==> Downloading https://linuxbrew.bintray.com/bottles-portable-ruby/portable-ruby-2.3.7.x86_64_linux.bottle.tar.gz

curl: (22) The requested URL returned error: 404 Not Found
==> Downloading https://github.com/Homebrew/homebrew-portable-ruby/releases/download/2.3.7/portable-ruby-2.3.7.x86_64_linux.bottle.tar.gz
```

jonchang commented:
Portable ruby might need to be special cased since it's hosted on homebrew.bintray.com and not linuxbrew.bintray.com

iMichka commented:
`unzip` is needed at two places in the new unpack_strategy code. On Linux unzip may not be there, so we need to install it before anything. There is some code in the dependency collector checking for `unzip`, but we do not have access to the dependency collector in unpack_strategy.

This is why I propose this change: https://github.com/Linuxbrew/brew/pull/749/commits/1d39b13aa45c511ef07c38ad406203a0629cf654

I have no clue if this is the right place for that kind of hack :)

jonchang commented:
suggest something like:

```
safe_system HOMEBREW_BREW_FILE, "install", "unzip"
```

along the lines of  `ensure_git_installed!` in `utils/git.rb`. 

is it possible to do this in `unpack_strategy.rb`?

iMichka commented:
Would we not need an `utils/unzip.rb` file with `ensure_unzip_installed`?

Also, isn't `ensure_git_installed` only installed when `brew tap` is called?

We could trigger this in the `initialize` from `UnpackStrategy`. But that code path will be hit each time that class is initialised, not sure if this will create a performance hit. Depends on the cost of checking if `unzip` is installed. Maybe there is a way to run it only once in that class, and store it in a class variable?

iMichka commented:
BTW calling `unzip` directly does not work ... even if installed by a formula. I tried to use `Formula["unzip"].opt_bin/"unzip"`, but this end in: `No available formula with the name "unzip"` ...

jonchang commented:
It would probably be necessary on some level to write an `ensure_formula_installed!` utility since it's already needed for `git`, `patchelf`, etc.

jonchang commented:
I'm looking at the code for unpack_strategy and dependency_collector and I'm not sure it's right. Where are you hitting the situation where unzip needs to be installed to unpack a download but depndency_collector doesn't properly install unzip beforehand?

jonchang commented:
Ok, I see the issue. I think changing the test to use `stub_formula_loader` before the zip and jar unpack strategy tests will fix it.

jonchang commented:
Giving up on this for the night, I tried a few different changes to the tests and couldn't get anything to work. Will try again tomorrow.

iMichka commented:
The test is one thing, the usage of brew is another one. You may install unzip only for the tests, but that will not fix the issue.

We may go for a quick solution like proposed by my 2 commits to get this going. I have not been merging homebrew-core since 2 days due to this merge missing, I do not want to have 1 week delay, which will be hard to catch up.

iMichka committed:
Use the mac domain to download the portable ruby bottles for Linux

Quick fix to get this going.

iMichka commented:
I fixed the bottle download problem. It was not really a problem as the mirror was being used. My solution is probably a quick fix but works.

iMichka committed:
Fix dependency cycle when install patchelf

Fixes:
==> Installing dependencies for unzip: patchelf, bzip2
==> Installing unzip dependency: patchelf
==> Installing dependencies for patchelf: patchelf
==> Installing patchelf dependency: patchelf
==> Installing dependencies for patchelf: patchelf
==> Installing patchelf dependency: patchelf
...

maxim-belkin commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -572,6 +572,7 @@ def inherited_options_for(dep)
   end
 
   def install_dependencies(deps)
+    return if formula.full_name == "patchelf" unless OS.mac?
```
I'd suggest checking if `formula.full_name` equals to `dep.to_formula.full_name` (on line 581 below) and skipping that dep in such a case.... and submitting this upstream.

maxim-belkin replied:
This would solve circular dependencies once and for all 🤺

iMichka replied:
Yes, why not. Not sure this is the best solution, because I actually not fully understand how the dependency cycle is happening. Maybe patchelf should not be added as a patchelf dependency at a first place. So not sure we are fixing this at the right place. But that should at least get us started.

maxim-belkin replied:
`compute_dependencies` puts together all dependencies and requirements. You could track down which function introduces `patchelf` as a dependency for `patchelf` and fix it, if you have the time...

iMichka replied:
I'll keep it in mind, I want 1.6.15 out as soon possible to be able to carry on with the core merges, and preparer 1.7.0.

iMichka commented:
Ok so my unzip solution works locally. The only problem I have now is that `HOMEBREW_UNZIP` is not correctly set while the tests are run on circle. I need to set that value when brew-testbot is run.

maxim-belkin commented on Library/Homebrew/cmd/update.sh:
```diff
@@ -383,6 +383,12 @@ user account:
   fi
 
+  # Needed for multiple unpack stategies
+  if [[ ! -x "$(command -v unzip)" ]]
```
why not simply...
```
! unzip &>/dev/null && brew install unzip
```
?

iMichka replied:
Because I was not aware of that syntax. I changed for your shorter version.

iMichka commented:
Ok got it to work ... but not really.

The tests are called with:
```
Homebrew/homebrew-test-bot 
ARGV: --ci-auto --no-pull --ci-pr --cleanup --test-default-formula --junit
```

I think cleanup is removing the unzip formula that was installed. This results in:
```
No such file or directory - /home/linuxbrew/.linuxbrew/opt/unzip/bin/unzip
```

(But I was quite happy to be able to set the path correctly for the tests :))

iMichka committed:
Use system unzip and subversion for the tests

iMichka commented:
@reitermarkus I see you worked on the unpacking strategy code.
One drawback of the current 1.6.15 code is that it needs `unzip` and `svn` at runtime. On Linux these are sometimes not installed. At first I thought I would just install `unzip` for all our users, using our bottle, I gave up because the I saw how many dependencies `svn` needs. So I won't install these dependencies.

I am just installing these in our CI so that the test pass, and so that I can tag the release. If we need these in linuxbrew-core, I'll install the system `unzip` and `svn`.

Do you think there would be a way to avoid having to call both of these?

iMichka commented:
@maxim-belkin @jonchang This is ready to be reviewed. I made 3 commits on top of the branch to fix the tests.

`unzip` and `svn` will be hard dependencies for Linuxbrew, until we find a better solution.

maxim-belkin commented on Library/Homebrew/cmd/vendor-install.sh:
```diff
@@ -21,12 +21,12 @@ elif [[ -n "$HOMEBREW_LINUX" ]]
   case "$HOMEBREW_PROCESSOR" in
     x86_64)
-      ruby_URL="$HOMEBREW_BOTTLE_DOMAIN/bottles-portable-ruby/portable-ruby-2.3.7.x86_64_linux.bottle.tar.gz"
+      ruby_URL="$HOMEBREW_BOTTLE_DEFAULT_DOMAIN_MACOS/bottles-portable-ruby/portable-ruby-2.3.7.x86_64_linux.bottle.tar.gz"
```
👍

maxim-belkin commented on Library/Homebrew/cmd/vendor-install.sh:
```diff
@@ -21,12 +21,12 @@ elif [[ -n "$HOMEBREW_LINUX" ]]
       ;;
     armv[67]*)
-      ruby_URL="$HOMEBREW_BOTTLE_DOMAIN/bottles-portable-ruby/portable-ruby-2.3.3_2.armv6_linux.bottle.tar.gz"
+      ruby_URL="$HOMEBREW_BOTTLE_DEFAULT_DOMAIN_MACOS/bottles-portable-ruby/portable-ruby-2.3.3_2.armv6_linux.bottle.tar.gz"
```
👍

maxim-belkin commented on Library/Homebrew/formula_installer.rb:
```diff
@@ -577,7 +577,11 @@ def install_dependencies(deps)
+        # Fix for patchelf
+        next if formula.full_name == dep.to_formula.full_name
+        install_dependency(dep, options)
+      end
```
Let's ask Homebrew if they are interested in this patch as well.

iMichka replied:
I propose to wait until 1.7.0 is merged, see if this still happens, and I'll open a PR for that. Are you OK with that?

maxim-belkin replied:
it doesn't hurt to ask... or does it? 😆

iMichka replied:
It's just that I do not want to open a PR if this is fixed with whatever other magic has happened in the code since then. I want to avoid unnecessary discussions. I promise I'll open a PR as soon as possible :)

maxim-belkin commented on .circleci/config.yml:
```diff
@@ -7,6 +7,9 @@ jobs:
     steps:
+      - run:
+         name: Install unzip and svn for tests
+         command: sudo apt-get update && sudo apt-get install -y unzip subversion
```
So, all tests in Homebrew assume presence of unzip and subversion?

iMichka replied:
Not only the tests. The code assumes it too. Basically all the calls to `unzip` and `svn` in unpack_strategies.rb.

maxim-belkin replied:
wow, ok. We should then indeed ask for their advice on how to handle this on Linux without unzip/svn. Requirements maybe?

iMichka replied:
> I think the unzip and svn problem was there before too. That code was in download_strategy.rb. So it was probably already broken. We just never noticed as there was no tests for that.

Yes, probably. Or vendoring. I think if the bug was there before, and nobody really complained, then it is probably not so urgent.
It will only bite those who build from source I guess.

jonchang replied:
Wouldn't dependency collector add unzip or svn as a dependency on formula install?

iMichka replied:
Right, actually it does ... so nothing to fix here.

It is just the tests that are broken, because the dependency collector does not kick in for that case.

jonchang replied:
I was looking into a way to install from bottle just for the tests but I didn't get very far. When I finish it I'll probably send it upstream instead of here.

iMichka replied:
You can do it, I tested it. But the test bot runs with --clean, which removes the bottle :D So this is not a good strategy.

jonchang replied:
An alternative is to skip the test if /usr/bin/unzip or whatever isn't detected.

iMichka commented:
I think the `unzip` and `svn` problem was there before too. That code was in download_strategy.rb. So it was probably already broken. We just never noticed as there was no tests for that.

iMichka commented:
Conclusion of all this (thanks for @jonchang to point me in the right direction): when needed, the dependency collector will install `unzip` or `svn`.

So the current fix to add  `unzip` or `svn` for the tests is fine. Maybe a refactoring of the code could avoid that though.

reitermarkus commented:
If you don't want the `unzip` and `svn` dependencies, you can add `:needs_macos` to the respective tests.
